[FLINK-24928][flink-runtime-web] Typing improvements for Flink UI

Airblader opened the pull request:
## What is the purpose of the change

This is a few commits which aim to slowly improve the code quality in Flink UI by improving type safety and various smaller fixes. With the exception of the last commit all changes are non-functional, and the last commit only fixes a bug which causes an error in the browser console.

This PR also adds `angular-testing-library` which will make writing tests in the future easier.

## Brief change log

See individual commits.

## Verifying this change

Flink UI doesn't yet have tests. I added a couple here due to the addition of `angular-testing-library`, but we need to add more tests, of course.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): yes
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: no
  - The serializers: no
  - The runtime per-record code paths (performance sensitive): no
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: no
  - The S3 file system connector: no

## Documentation

  - Does this pull request introduce a new feature? no
  - If yes, how is the feature documented? not applicable

nicoweidner commented on flink-runtime-web/web-dashboard/src/app/interfaces/job-detail.ts:
```diff
@@ -149,12 +147,12 @@ export interface NodesItemLinkInterface {
 }
 
-export interface JobDetailCorrectInterface extends JobDetailInterface {
+export interface JobDetailCorrect extends JobDetail {
```
I am rather amused by interface names such as this one...

nicoweidner commented on flink-runtime-web/web-dashboard/src/app/interfaces/job-checkpoint.ts:
```diff
@@ -113,7 +111,7 @@ export interface CheckPointTaskStatistics {
 
 export interface CheckPointConfig {
-  mode: SafeAny;
+  mode: 'exactly_once' | string;
```
Why the special role for `exactly_once`?

Airblader replied:
That's the only value I could find in the code, but I assume there's more, so I documented the known value and played it safe (IDEs can still auto-complete that value).

nicoweidner replied:
There seems to be only this and `at_least_once` in Flink, but yeah, not for this PR to resolve...

nicoweidner commented on flink-runtime-web/web-dashboard/src/app/pages/job/timeline/job-timeline.component.ts:
```diff
@@ -143,16 +189,18 @@ export class JobTimelineComponent implements AfterViewInit, OnDestroy {
+    this.mainChartInstance.on('click', (e: { x: number; y: number }) => {
       if (this.mainChartInstance.getSnapRecords(e).length) {
-        const data = (this.mainChartInstance.getSnapRecords(e)[0] as SafeAny)._origin;
+        const data = ((this.mainChartInstance.getSnapRecords(e)[0] as unknown) as {
```
... can we really not avoid typing this as `number[0]` and then later on typecasting it to access `._origin` after all?

Airblader replied:
`… as number as { _origin: }` will not work, no. The whole reason `as unknown` is in here in the first place is because TS will not let you type-assert a known type into another one when there is no overlap and it could never work. It then recommends the only way around that, which is asserting it to `unknown` first.

Really this is all rather "broken", obviously the types do not match what is going on here. But I haven't (yet) addressed that.

nicoweidner replied:
Sorry, my message was a bit garbled. What I meant to point out was that it's rather strange that we apparently have a `number[]`, but then suddenly access property `._origin`. But you already addressed that in your answer

nicoweidner commented on flink-runtime-web/web-dashboard/src/app/share/common/dagre/graph.ts:
```diff
@@ -40,10 +40,12 @@ export interface LayoutNodeOptions {
+  [key: string]: unknown;
 
-  detail: SafeAny;
+  detail: {
+    [key: string]: unknown;
+  };
```
Isn't this just `Record<string, unknown>`?

Airblader replied:
It is! Will change it.

nicoweidner commented on flink-runtime-web/web-dashboard/src/app/share/common/dagre/svg-container.component.ts:
```diff
@@ -32,7 +32,8 @@ import * as d3 from 'd3';
 
-import { SafeAny } from 'interfaces';
+/* eslint-disable-next-line @typescript-eslint/no-explicit-any */
+export type SafeAny = any;
```
No way to remove this?

Airblader replied:
We don't need to export this type, though, that I will change.

nicoweidner commented on flink-runtime-web/web-dashboard/src/app/share/common/dagre/svg-container.component.ts:
```diff
@@ -94,12 +86,12 @@ export class SvgContainerComponent implements OnInit, AfterContentInit {
         }
         this.zoomEvent.emit(this.zoom);
-        this.transformEvent.emit(this.containerTransform as SafeAny);
+        this.transformEvent.emit((this.containerTransform as unknown) as { x: number; y: number; scale: number });
```
Why the weird double casting? I looked this one up as an example (since there were so many cases), and `containerTransform` has properties `x`, `y`, `k`. The latter might mean the same as `scale`, but here `scale` will just be undefined, right?

Airblader replied:
Yes, so there is some really weird type mismatch going on between `k` and `scale` being used. You cannot type-assert this directly, the compiler won't let you because there is not enough overlap for that to ever work. The trick to assert it to `unknown` first is recommended by the compiler itself. This is definitely a symptom of something being broken, though.

nicoweidner replied:
Ok, I was just confused by the weird type mismatches going on, but as you said - let's fix it some other time

Airblader commented on flink-runtime-web/web-dashboard/src/app/pages/job/checkpoints/subtask/job-checkpoints-subtask.component.ts:
```diff
@@ -40,31 +55,42 @@ import { deepFind } from 'utils';
+  public readonly sortAckTimestampFn = createSortFn(item => item.ack_timestamp);
+  public readonly sortEndToEndDurationFn = createSortFn(item => item.end_to_end_duration);
+  public readonly sortStateSizeFn = createSortFn(item => item.state_size);
+  public readonly sortCpSyncFn = createSortFn(item => item.checkpoint?.sync);
```
The sort functions here never return 0, so there is no stability, no. But that's how it was before, I just switched from a type-unsafe string path to a type-safe selector function.

Airblader commented on flink-runtime-web/web-dashboard/src/app/pages/job/overview/list/job-overview-list.component.ts:
```diff
@@ -16,12 +16,17 @@
+function createSortFn(
+  selector: (item: NodesItemCorrectInterface) => number | string | undefined
+): NzTableSortFn<NodesItemCorrectInterface> {
+  return (pre, next) => (selector(pre)! > selector(next)! ? 1 : -1);
```
Yes, we are. Note that `undefined > …`, `… > undefined` and `undefined > undefined` are all "valid" in the sense that they don't error, so we're not running into the problem of possibly causing errors here (the behavior was pre-existent anyway, now it's just visible).

nicoweidner replied:
Maybe we could just return -1 whenever any of the values is undefined? That should have the same effect and be (a little bit) less awkward

Airblader replied:
No need to – a comparison with `undefined` on either side will always yield `false` (and thus lead to `-1`) anyway.

nicoweidner replied:
Yes, that's why I suggested it. I just wanted to get rid of the assertions ;-)

Airblader commented on flink-runtime-web/web-dashboard/src/app/pages/submit/submit.component.ts:
```diff
@@ -71,24 +113,14 @@ export class SubmitComponent implements OnInit, OnDestroy {
     );
   }
 
-  /**
-   * Delete jar
-   *
-   * @param jar
-   */
```
Would you like me to add them back? :-)

nicoweidner replied:
Only if you add these lines as well:
```
     * @return void
     * @see JarFilesItem
```

Airblader commented on flink-runtime-web/web-dashboard/src/app/interfaces/job-checkpoint.ts:
```diff
@@ -18,7 +18,7 @@
 import { SafeAny } from './safe-any';
 
-export interface CheckPointInterface {
+export interface CheckPoint {
```
Sure, done.

Airblader commented:
@nicoweidner Thanks for the review, I addressed all your feedback, I believe. Note that I'm fully aware that this PR doesn't make things perfect – it'll take several iterations. Right now I focused on changes that do not cause behavioral changes but can be applied across the entire application. We'll of course have to go over things in more detail in later steps and sort more things out, then ideally by adding tests as well.

nicoweidner reviewed (approved):
Quite a lot of good typing improvements, thanks for wading through this!

AHeise commented:
Thanks for the contribution. I cannot access the content but looking at the size, it's safe to assume that we are not talking about a hotfix. Please create a Jira ticket and use that instead of hotfix. The ticket will be then a point where future developers can refer to. (Imagine you are introducing a bug, it would be much easier to reference that in Jira as caused by)

Airblader commented:
@AHeise Would it be OK to create an issue and put multiple PRs on it? It'll be rather weird to create random open-ended issues that are basically all the same, so I'm wondering what the best approach for that would be. If you prefer individual issues I'm happy to do that too, though.

AHeise commented:
It's in general okay to have multiple PRs for one issue. 
However, at this size, I'd rather use an umbrella ticket and have smaller tasks on it like ("Improve typing in Web UI", "Reduce code clutter in X", "Increase test coverage of B") each having their own ticket.

Airblader commented:
@AHeise Fair enough, thanks. I've created an umbrella and made this PR a sub-task thereof (https://issues.apache.org/jira/browse/FLINK-24928). I'm not sure if the JIRA will link up the PR with the issue after the fact now, though.

Airblader commented:
@AHeise And the CI has passed again now, too. :-)

MartijnVisser commented:
@AHeise Please merge :)

AHeise commented:
My pleasure. Thanks!
